Delegate bean resolution context when resolving proxy target

dstepanov opened the pull request:
Added a check not to cache proxies in `DefaultBeanContext` `scopedProxies` for `@Any` bean producers, not sure if we should cache anything as it can leak the instance for some short-lived scopes.

dstepanov committed:
Delegate bean resolution context when resolving non-lazy proxy target

jameskleeh commented on aop/src/main/java/io/micronaut/aop/chain/ConstructorInterceptorChain.java:
```diff
@@ -181,16 +181,16 @@ public T invoke(T instance, Object... arguments) {
                     originalParameters,
                     0,
-                    originalParameters.length - 3
+                    originalParameters.length - 4
```
How would this work with old style bean definitions? Couldn't this break with those definitions?

dstepanov committed:
Store the count of additional proxy constructor parameters

dstepanov committed:
Delegate bean resolution context when resolving lazy proxy target

dstepanov committed:
Store and reuse target for lazy beans

dstepanov commented:
I have added a way to pass the number of added constructor parameters in proxies that should solve the backward compatibility issue.

Implemented bean resolution context storing and propagating for lazy interceptor target beans.

Also added double-check locking to cache the target in lazy beans: 

```java
    public Object interceptedTarget() {
        ProxiedSomeBeanConsumer var1 = this.$target;
        if (var1 == null) {
            synchronized(this) {
                var1 = this.$target;
                if (var1 == null) {
                    this.$target = (ProxiedSomeBeanConsumer)((DefaultBeanContext)this.$beanLocator).getProxyTargetBean(this.$beanResolutionContext, Argument.of(ProxiedSomeBeanConsumer.class, $ProxiedSomeBeanConsumer$Definition$Intercepted$Definition$Reference.$ANNOTATION_METADATA, new Class[0]), this.$beanQualifier);
                    this.$beanResolutionContext = null;
                }
            }
        }

        return this.$target;
    }
 ```

graemerocher commented:
@dstepanov tests are failing, those need addressing before review

dstepanov committed:
Cache target only on demand

dstepanov commented:
It looks like caching target cannot be done by default.
I have added an option `@Around(cacheableLazyTarget)` to preserve the work I did

graemerocher commented on inject/src/main/java/io/micronaut/context/DefaultBeanContext.java:
```diff
@@ -4248,4 +4328,34 @@ public void close() {
         }
     }
+
+    private class DefaultBeanResolutionContext extends AbstractBeanResolutionContext {
```
Rename to `SingletonBeanResolutionContext` since there is already a `DefaultBeanResolutionContext` in `io.micronaut.context`

dstepanov committed:
Address review
